### ISSUE TITLE
Add mypy to typecheck the codebase

### ISSUE THREAD

**Issue by nicarl** [joke2k/faker#1334]
* Faker version: master
* OS: all

How about adding mypy (http://mypy-lang.org/) to typecheck the codebase? 
One could start to add type annotations and once the codebase is covered activate mypy type checking in the CI.

**Comment by UrielMaD** [comment]
Hello @nicarl, can I help on typechek the codebase?

**Comment by nicarl** [comment]
Hey that would be great! At the moment I am using the following mypy.ini and mypy 0.790:

```
[mypy]
python_version = 3.8
disallow_subclassing_any = True
disallow_untyped_calls = True
disallow_untyped_defs = True
disallow_incomplete_defs = True
check_untyped_defs = True
warn_redundant_casts = True
warn_unused_ignores = True
warn_unreachable = True
allow_redefinition = True
pretty = True
follow_imports=silent 
```

I think best is to go file by file. The root folder (containing `config.py`, `cli.py` ...) is a good place to start I think.

Should we check in the mypy config and already add the dependency?

**Comment by UrielMaD** [comment]
@nicarl sure! I'm kind of new contributing and I just started to using mypy, I saw that you just merged some files, so, could I start right into the root folder?

**Comment by nicarl** [comment]
Yes you can start in the root folder. You can just ping me if you get stuck.

**Comment by UrielMaD** [comment]
@nicarl Hello again Nicarl, I just did a little implementation of type checking in the cli.py file, but I realize that this topic is deeper than I thought, I'm very interested on dive into this and I was wondering that I could need a little guidence on this

**Comment by mdantonio** [comment]
Hello,
just as information: mypy ignores type hints from packages that do not include a py.typed marker

https://www.python.org/dev/peps/pep-0561/#packaging-type-information

_Package maintainers who wish to support type checking of their code MUST add a marker file named py.typed to their package supporting typing._

If some type hints have been added (even if very partial), would be useful to include the py.typed file to the release to let users to start using type hints and encourage contributions

**Comment by fcurella** [comment]
Thank you @mdantonio , I've added the `py.typed` file in https://github.com/joke2k/faker/commit/52e5018e3fcf5d2b176d2031672c56bc2140ebc9

**Comment by mdantonio** [comment]
Great, I will test it as soon as a new version will be released

**Comment by sirxbiff** [comment]
Now the `py.typed` file is in there mypy complains about `error: Call to untyped function "Faker" in typed context` for a simple line like:
```python
fake = Faker()
```
Adding type annotations to the line doesn't help. Removing the `py.typed` file manually from the package does fix the issue. 

I think `py.typed` should only be added after more type annotations have been added.

**Comment by nicarl** [comment]
I started to add type annotations some time ago and still have a branch with more type annotations lying around. I will try to rebase it and contribute what I have at the moment.
Any help with adding type annotations even for small parts of the code is highly appreciated

**Comment by nicarl** [comment]
My current branch can be found here: https://github.com/nicarl/faker/tree/add_type_annotations_for_providers
